File: frontend/lib/services/firebase/firestore_service.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:hackncsu_today/exception.dart';
import 'package:hackncsu_today/models/event/event_data.dart';
import 'package:hackncsu_today/models/event/event_state.dart';
import 'package:hackncsu_today/models/hack_user.dart';
import 'package:hackncsu_today/models/team.dart';
import 'package:hackncsu_today/services/firebase/functions_service.dart';
import 'package:riverpod_annotation/riverpod_annotation.dart';

part 'firestore_service.g.dart';

@riverpod
FirebaseFirestoreService firebaseFirestoreService(Ref ref) {
  return FirebaseFirestoreService();
}

/// Note: firestore rules are set up such that authenticated users may only
/// get/update their own user data. Cloud functions can be used to
/// get a small amount of data about other users
class FirebaseFirestoreService {
  final FirebaseFirestore _firestore = FirebaseFirestore.instance;

  static const String _usersCollection = 'users';
  static const String _eventCollection = 'event';
  static const String _teamsCollection = 'teams';

  static const String _eventStateDoc = 'state';
  static const String _eventDataDoc = 'data';

  /// creates a new user document in Firestore with the given data.
  /// if the user already exists, it returns the existing user document.
  /// this is meant for use in the login flow
  Future<HackUser> createUser(AuthenticateFunctionResponse response) async {
    final userDocRef = _firestore.collection(_usersCollection).doc(response.id);
    final docSnapshot = await userDocRef.get();

    if (docSnapshot.exists) {
      final existingUser = HackUser.fromJson(docSnapshot.data()!);

      // Skip this check in debug mode to allow for flexible testing with different user types
      if (!kDebugMode) {
        if (existingUser is Organizer && !response.isOrganizer) {
          throw FirebaseFirestoreException(
            'User with ID ${response.id} is an organizer, but previous records indicate they are not.',
          );
        } else if (existingUser is Participant && response.isOrganizer) {
          throw FirebaseFirestoreException(
            'User with ID ${response.id} is a participant, but previous records indicate they are an organizer.',
          );
        }
      }

      return existingUser;
    } else {
      final HackUser hackUser;

      if (response.isOrganizer) {
        hackUser = Organizer(id: response.id);
      } else {
        hackUser = Participant(
          id: response.id,
          firstName: response.firstName!,
          lastName: response.lastName!,
          phone: response.phone,
          email: response.email,
          dietaryRestrictions:
              [], // TODO: add ability to get dietary restrictions
          shirtSize: null,
          eventsAttended: [],
          hadFirstLunch: false,
          hadDinner: false,
          hadBreakfast: false,
          hadSecondLunch: false,
        );
      }

      await userDocRef.set(hackUser.toJson()).catchError((error) {
        throw FirebaseFirestoreException('Failed to create user data: $error');
      });
      return hackUser;
    }
  }

  /// updates the user document in Firestore with the given user data.
  Future<void> updateUser(HackUser user) async {
    _firestore
        .collection(_usersCollection)
        .doc(user.id)
        .set(user.toJson(), SetOptions(merge: true))
        .catchError((error) {
          throw FirebaseFirestoreException(
            'Failed to update user data: $error',
          );
        });
  }

  /// fetches the user data from Firestore for the given user ID.
  /// returns null if the user does not exist.
  /// NOTE: this will fail for participants if they try to fetch another user's data
  /// because of Firestore rules. they can only fetch their own data.
  /// or use cloud functions to fetch limited data about other users.
  Future<HackUser?> fetchUser(String userId) async {
    final userDocRef = _firestore.collection(_usersCollection).doc(userId);
    final docSnapshot = await userDocRef.get().catchError((error) {
      throw FirebaseFirestoreException('Failed to fetch user data: $error');
    });

    if (docSnapshot.exists) {
      return HackUser.fromJson(docSnapshot.data()!);
    } else {
      return null;
    }
  }

  /// Streams user data from Firestore for the given user ID.
  Stream<HackUser?> streamUser(String userId) {
    final userDocRef = _firestore.collection(_usersCollection).doc(userId);
    return userDocRef.snapshots().map((snapshot) {
      if (snapshot.exists && snapshot.data() != null) {
        return HackUser.fromJson(snapshot.data()!);
      }
      return null;
    });
  }

  /// fetches the team data from Firestore for the given team ID.
  /// returns null if the team does not exist.
  /// NOTE: this will fail for participants if they try to fetch another team's data.
  Future<Team?> fetchTeamData(String teamId) async {
    final teamDocRef = _firestore.collection(_teamsCollection).doc(teamId);
    final docSnapshot = await teamDocRef.get().catchError((error) {
      throw FirebaseFirestoreException('Failed to fetch team data: $error');
    });

    if (docSnapshot.exists) {
      return Team.fromJson(docSnapshot.data()!);
    } else {
      return null;
    }
  }

  /// Streams event data from Firestore.
  Stream<EventData?> streamEventData() {
    final eventDataRef = _firestore
        .collection(_eventCollection)
        .doc(_eventDataDoc);

    return eventDataRef.snapshots().map((snapshot) {
      if (snapshot.exists && snapshot.data() != null) {
        return EventData.fromJson(snapshot.data()!);
      }
      return null;
    });
  }

  /// Streams event state from Firestore
  Stream<EventState?> streamEventState() {
    final eventStateRef = _firestore
        .collection(_eventCollection)
        .doc(_eventStateDoc);

    return eventStateRef.snapshots().map((snapshot) {
      if (snapshot.exists && snapshot.data() != null) {
        return EventState.fromJson(snapshot.data()!);
      }
      return null;
    });
  }

  // Below are functions that are only authorized for organizers
  // They will fail for participants by nature of our Firestore rules

  /// initializes event in Firestore.
  /// this can be called from the admin panel to set up the event
  /// it should be called once at the start of the event
  /// and then we can update the relevant data as needed
  Future<void> initializeEvent() async {
    final eventData = EventData(
      tracks: [
        'Track 1',
        'Track 2',
        'Track 3',
      ],
      externalResources: [
        Resource.link('Hack_NCState Website', 'https://hackncstate.org'),
        Resource.link(
          'Centennial Map',
          'https://maps.ncsu.edu/#/buildings/783A',
        ),
      ],
      internalResources: [
        Resource.link('Discord Server', 'https://example.com', hidden: true),
        Resource.link('Schedule', 'https://example.com', hidden: true),
        Resource.link(
          'Opening Ceremony Slides',
          'https://hackncstate.org',
          hidden: true,
        ),
        Resource.action('Catering Options', ActionType.menu, hidden: true),
      ],
    );

    final stateRef = _firestore
        .collection(_eventCollection)
        .doc(_eventStateDoc);
    final dataRef = _firestore.collection(_eventCollection).doc(_eventDataDoc);

    final batch = _firestore.batch();

    batch.set(stateRef, EventState.initial().toJson());
    batch.set(dataRef, eventData.toJson());

    batch.commit().catchError((error) {
      throw FirebaseFirestoreException(
        'Failed to initialize event data: $error',
      );
    });
  }

  Future<void> updateEventData(EventData eventData) async {
    final dataRef = _firestore.collection(_eventCollection).doc(_eventDataDoc);

    await dataRef.set(eventData.toJson(), SetOptions(merge: true)).catchError((
      error,
    ) {
      throw FirebaseFirestoreException('Failed to update event data: $error');
    });
  }

  Future<void> updateEventState(EventState eventState) async {
    final stateRef = _firestore
        .collection(_eventCollection)
        .doc(_eventStateDoc);

    await stateRef.set(eventState.toJson(), SetOptions(merge: true)).catchError(
      (error) {
        throw FirebaseFirestoreException(
          'Failed to update event state: $error',
        );
      },
    );
  }

  // Below are functions that are used only during debugging

  /// Sets the user type (useful for debugging purposes).
  Future<void> debugSetUserType(String id, String type) async {
    if (kDebugMode) {
      _firestore
          .collection(_usersCollection)
          .doc(id)
          .update({'type': type})
          .catchError((error) {
            throw FirebaseFirestoreException('Failed to switch view: $error');
          });
    }
  }
}

class FirebaseFirestoreException extends AppException {
  const FirebaseFirestoreException(super.message);
}

File: src/services/firestore.service.ts
import { firestore } from "@/lib/firebase-config";
import { EventConfigSchema, type EventConfig } from "@/types/event";
import { TeamSchema, type Team } from "@/types/team";
import { UserSchema, type UserData } from "@/types/user";
import {
	collection,
	deleteDoc,
	doc,
	getDoc,
	getDocs,
	onSnapshot,
	setDoc,
	updateDoc,
} from "firebase/firestore";

const collections = {
	users: "users",
	event: "event",
	teams: "teams",
};

export const firestoreService = {
	// this will fail if the user is a participant
	// and they try accessing other user's data
	fetchUser: async (userId: string) => {
		const userDocRef = doc(firestore, collections.users, userId);
		const userSnapshot = await getDoc(userDocRef);

		if (userSnapshot.exists()) {
			return UserSchema.parse(userSnapshot.data());
		}

		return null;
	},

	onUserSnapshot: (
		userId: string,
		callback: (data: UserData | null) => void,
	) => {
		const userDocRef = doc(firestore, collections.users, userId);
		return onSnapshot(userDocRef, (snapshot) => {
			if (snapshot.exists()) {
				const data = UserSchema.parse(snapshot.data());
				callback(data);
			} else {
				callback(null);
			}
		});
	},

	updateUser: async (userId: string, data: Partial<UserData>) => {
		const userDocRef = doc(firestore, collections.users, userId);
		await updateDoc(userDocRef, data);
	},

	fetchTeam: async (teamId: string) => {
		const teamDocRef = doc(firestore, collections.teams, teamId);
		const teamSnapshot = await getDoc(teamDocRef);

		if (teamSnapshot.exists()) {
			return TeamSchema.parse(teamSnapshot.data());
		}
		return null;
	},

	fetchAllTeams: async () => {
		const teamsCollectionRef = collection(firestore, collections.teams);
		const docs = await getDocs(teamsCollectionRef);

		return docs.docs.map((doc) => TeamSchema.parse(doc.data()));
	},

	fetchAllUsers: async () => {
		const usersCollectionRef = collection(firestore, collections.users);
		const docs = await getDocs(usersCollectionRef);

		return docs.docs.map((doc) => UserSchema.parse(doc.data()));
	},

	deleteTeam: async (teamId: string) => {
		const team = await firestoreService.fetchTeam(teamId);

		const teamDocRef = doc(firestore, collections.teams, teamId);

		await deleteDoc(teamDocRef);

		if (team) {
			await firestoreService.updateUser(team.creatorId, { teamId: null });

			if (team.status === "approved") {
				// remove teamId from all members if approved (hence all members are associated, not just the creator)
				for (const memberId of team.memberIds) {
					await firestoreService.updateUser(memberId, { teamId: null });
				}
			}
		}
	},

	updateTeam: async (teamId: string, data: Partial<Team>) => {
		const teamDocRef = doc(firestore, collections.teams, teamId);
		await updateDoc(teamDocRef, data);
	},

	onTeamSnapshot: (teamId: string, callback: (data: Team | null) => void) => {
		const teamDocRef = doc(firestore, collections.teams, teamId);
		return onSnapshot(teamDocRef, (snapshot) => {
			if (snapshot.exists()) {
				const data = TeamSchema.parse(snapshot.data());
				callback(data);
			} else {
				callback(null);
			}
		});
	},

	onEventConfigSnapshot: (callback: (config: EventConfig | null) => void) => {
		const eventDocRef = doc(firestore, collections.event, "main");

		return onSnapshot(eventDocRef, (snapshot) => {
			if (snapshot.exists()) {
				const data = EventConfigSchema.parse(snapshot.data());
				callback(data);
			} else {
				callback(null);
			}
		});
	},

	fetchEventConfig: async () => {
		const eventDocRef = doc(firestore, collections.event, "main");
		const eventSnapshot = await getDoc(eventDocRef);

		if (eventSnapshot.exists()) {
			return EventConfigSchema.parse(eventSnapshot.data());
		}

		return null;
	},

	updateEventConfig: async (data: Partial<EventConfig>) => {
		const eventDocRef = doc(firestore, collections.event, "main");
		await setDoc(eventDocRef, data, { merge: true });
	},

	clearEventConfig: async () => {
		const eventDocRef = doc(firestore, collections.event, "main");
		await deleteDoc(eventDocRef);
	},

	// this only works in debug mode. useful for testing different user roles
	debugSetUserType: async (
		userId: string,
		type: "organizer" | "participant",
	) => {
		if (import.meta.env.DEV) {
			const userDocRef = doc(firestore, collections.users, userId);
			const userSnapshot = await getDoc(userDocRef);

			if (userSnapshot.exists()) {
				const updatedData = {
					role: type,
				};
				await updateDoc(userDocRef, updatedData);
			}
		}
	},
	// for creating example users in debug mode
	debugCreateSampleParticipants: async () => {
		if (import.meta.env.DEV) {
			const sampleUsers = Array.from({ length: 10 }).map((_, i) => ({
				id: `sample-user-${i}`,
				username: `sampleuser${i}`,
				role: "participant" as const,
				email: `sampleuser${i}@example.com`,
				firstName: `Sample`,
				lastName: `User ${i}`,
				phone: "123-456-7890",
				shirtSize: "M",
				dietaryRestrictions: "None",
				rfidUUID: `rfid-${i}`,
				attendedEvents: [],
				hadFirstLunch: false,
				hadSecondLunch: false,
				hadBreakfast: false,
				hadDinner: false,
			}));

			for (const user of sampleUsers) {
				const userDocRef = doc(firestore, collections.users, user.id);
				await setDoc(userDocRef, user);
			}
		}
	},
};
